### PR TITLE
fix(parser): scanner drops every entity when HEADER FILE_NAME contains '#' (#654)

### LINKED ISSUE #654: Model Parsing and Viewing Abnormalities
Bug:
This model was created with CATIA. It displays normally with correct colors in BIMVision. However, it loses material and color rendering in webifc, while IFCLite Viewer can only show property information without any geometric display. Could you please help check this issue? Thank you very much.

Test Model:
[CatiaBuilding.zip](https://github.com/user-attachments/files/27645420/CatiaBuilding.zip)

Viewing effect with IFCLite Viewer:

Image

viewing effect with webifc (via https://ifc.babylonpress.org/):

Image

viewing effect with BIMVision:

Image

### PR DESCRIPTION
**Pull request by louistrue**
## Summary

Two commits:

1. **`fix(parser): scanner drops every entity when HEADER FILE_NAME contains '#'`** — the actual bug fix.
2. **`build: rebuild WASM from Rust on every deploy; remove committed binaries`** — closes the structural gap that made #654 hard to ship (and at one point caused a wrong-bytes regression mid-PR).

## The parser bug

CATIA emits a `FILE_NAME` whose first argument is an encoded filename containing a literal `#` inside the quoted string (`'…\X0\2#.ifc'`). `EntityScanner::next_entity` anchored on `memchr(b'#')` without skipping the HEADER section or validating the candidate, latched onto that in-string `#`, then called `find_entity_end` starting *inside* an open STEP string with `in_string = false`. Quote parity inverted for every subsequent `'`; every `;` for the rest of the file looked "inside a string" and was skipped. The scanner returned zero entities — so IFCLite Viewer showed properties (different parser path) but rendered no geometry.

Fixed by skipping past `DATA;` (quote-aware, ISO 10303-21 `''` aware) and validating `#`+digit before treating a candidate as an entity start. Matches what `build_entity_index` and the TS `StepTokenizer` already do. Three regression tests cover the CATIA pattern, `DATA;` inside a `FILE_DESCRIPTION`, and an escaped-quote-pair edge case.

End-to-end against the issue's `CatiaBuilding.ifc`:

| | Before | After |
|---|---:|---:|
| Entities scanned | 0 | 83 375 |
| Unit scale | 1.0 (default) | 0.001 (mm→m) |
| Walls with mesh | 0 / 770 | 769 / 770 |
| Doors / slabs / windows with mesh | 0 | all |

The one remaining empty wall has a NULL representation in the file itself. `IFCSITE` / `IFCCURTAINWALL` are spatial containers without direct geometry. The color-loss half of #654 (web-ifc not picking up `IfcIndexedColourMap`) is a separate bug not addressed here.

## The build pipeline

The original symptom for #654 was a stale `.wasm` bundle shipping to production. A maintainer changes Rust source, tests pass, the PR merges, but the viewer keeps loading whatever `.wasm` was committed last week because nothing in the pipeline rebuilds it. During this very PR we hit it twice — once when cargo's incremental cache reused stale `ifc-lite-core` and produced a wasm without the fix, and once when a merge-conflict resolution picked the wrong bytes for the committed binary.

The root issue is committing a derived artifact and trusting humans to commit correct bytes. After this PR:

- `packages/wasm*/pkg/ifc-lite_bg.wasm` (+ generated `.js`/`.d.ts`/snippets) is gitignored. The `pkg/` directory still ships in the npm publish surface; contents are regenerated by `scripts/build-wasm.sh` on every build.
- `packages/wasm/package.json` and `packages/wasm-threaded/package.json` gain a `build` script — until now turbo treated them as no-ops because they had no script.
- `turbo.json` adds Rust paths to `globalDependencies` (content-hash cache invalidation on Rust changes) and Rust env vars to `globalPassThroughEnv` (turbo strips them by default).
- `scripts/vercel-install.sh` + `scripts/vercel-build.sh` bootstrap a Rust toolchain on Vercel's iad1 runner. Both Vercel projects (`ifc-lite`, `ifc-lite-viewer-embed`) point at them.
- `.github/workflows/desktop-compat.yml` now installs Rust + wasm-pack so the desktop frontend build can rebuild wasm too.
- `README.md` updates the contributing instructions — contributors now need a Rust toolchain (one-time install).

**Cost on Vercel Pro:** JS-only deploys hit turbo's content cache and pay ~0 s extra. Rust-touching deploys add ~30–60 s warm / ~3–5 min cold. Realistic monthly load stays well under the plan ceiling.

## Reviewer feedback addressed

- **Codex `parser.rs`** — "Skip only the real DATA section marker" → quote-aware `data_section_start` with ISO 10303-21 escape handling + two regression tests.
- **CodeRabbit `parser.rs`** — same concern as Codex, same fix.
- **CodeRabbit `build-wasm.sh`** — `-x` (executable) replaces `-f` (file exists) on the `~/.cargo/bin/wasm-pack` probe.

## Test plan

- [x] `cargo test -p ifc-lite-core --lib parser::tests` — 14 tests, including 3 new regressions.
- [x] `cargo test -p ifc-lite-geometry --lib` — 163 tests.
- [x] `npx turbo build --filter=@ifc-lite/viewer` clean — 26/26 tasks.
- [x] Functional WASM test against CATIA file via Node harness: 83 375 entities, 1 765 geometry jobs.
- [ ] Vercel preview visual verification: load `CatiaBuilding.ifc` in the preview, confirm geometry renders. Preview URL: https://ifc-lite-git-fix-issue-654-catia-header-hash-ltplus.vercel.app/

Closes #654.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: rust/core/src/parser.rs [hunk 1]
```diff
@@ -516,6 +540,40 @@ impl<'a> EntityScanner<'a> {
...
+    let limit = len - MARKER.len();
+    let mut i = 0;
+    while i <= limit {
+        // Anchor on 'D' to keep the scan cheap.
+        let Some(off) = memchr::memchr(b'D', &bytes[i..=limit]) else {
+            return 0;
+        };
+        let p = i + off;
+        if &bytes[p..p + MARKER.len()] == MARKER {
+            return p + MARKER.len();
```
[the remark this answers is not in the document]
Comment (louistrue) [reply rust/core/src/parser.rs]: Good catch — addressed in 69088a3d. `data_section_start` now tracks single-quote state (with the ISO 10303-21 `''` escaped-pair rule) so the `DATA;` marker is only matched outside any quoted HEADER field. Added two regressions: `DATA;` literal in a `FILE_DESCRIPTION`, and an escaped-quote string that holds `DATA;` past a doubled-quote pair.

File: rust/core/src/parser.rs [hunk 2]
```diff
@@ -516,6 +540,40 @@ impl<'a> EntityScanner<'a> {
...
+        }
+        i = p + 1;
+    }
+    0
+}
```
[the remark this answers is not in the document]
Comment (louistrue) [reply rust/core/src/parser.rs:597]: Already addressed in commit 69088a3d — `data_section_start` is quote-aware (tracks single-quote state with the ISO 10303-21 `''` escaped-pair rule), so the `DATA;` marker is only matched outside any quoted HEADER field. Two regression tests cover it: `DATA;` literal inside a `FILE_DESCRIPTION`, and an escaped-quote string that contains `DATA;` past a doubled-quote pair. (CodeRabbit auto-resolution didn't pick this up on re-scan — manually marking resolved.)

File: scripts/build-wasm.sh [hunk 3]
```diff
@@ -17,16 +17,36 @@ echo "🦀 Building IFC-Lite WASM..."
...
+# touch Rust — should rebuild from source when possible but fall back to
+# the committed artifact when not. Hard-failing here would break every
+# `turbo build` on those hosts even though the .wasm they need is already
+# present in the repo.
 WASM_PACK="wasm-pack"
 if ! command -v wasm-pack &> /dev/null; then
-  # Try cargo's bin directory (common location for cargo-installed binaries)
   CARGO_BIN="$HOME/.cargo/bin/wasm-pack"
   if [ -f "$CARGO_BIN" ]; then
     WASM_PACK="$CARGO_BIN"
```
[the remark this answers is not in the document]
Comment (louistrue) [reply scripts/build-wasm.sh:36]: Good catch — fixed in the latest commit. `-x` (executable) replaces `-f` (file exists) so a non-exec file at `~/.cargo/bin/wasm-pack` no longer passes the guard.
